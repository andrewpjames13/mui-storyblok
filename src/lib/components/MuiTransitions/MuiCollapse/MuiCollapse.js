import React, { createElement } from 'react';
import PropTypes from 'prop-types';
import { Collapse } from '@material-ui/core';
import StoryBlok from '../../../utils/Storyblok';
import MuiGridList from '../../MuiGridList/MuiGridList';

const MuiCollapse = ({
  rootClass,
  collapsedHeight,
  transitionIn,
  timeOut,
  content,
}) => {
  const components = {
    MuiGridList,
  };

  const styles = StoryBlok.arrayToMuiStyles(rootClass);

  return (
    <div className={styles.root}>
      <Collapse
        collapsedHeight={collapsedHeight}
        in={transitionIn}
        timeOut={timeOut}
      >
        <div>
          {content.map((item, index) => createElement(
            components[item.component],
            Object.assign(item, { key: index }),
          ))}
        </div>
      </Collapse>
    </div>
  );
};

export default MuiCollapse;

MuiCollapse.propTypes = {
  /**
   * stroyblok multiselect of css classes
   * Mui Override or extend the styles applied to the component.
   */
  rootClass: PropTypes.arrayOf(PropTypes.string),
  /**
   * mui props: number | string
   * The height of the container when collapsed.
   */
  collapsedHeight: PropTypes.string,
  /**
   * mui props: number | string
   * The duration for the transition, in milliseconds.
   * You may specify a single timeout for all transitions, or individually with an object.
   * Set to 'auto' to automatically calculate transition time based on height.
   */
  timeOut: PropTypes.string,
  /**
   * mui props: bool
   * If true, the component will transition in.
   */
  transitionIn: PropTypes.bool,
  /**
   * Content passed to render
   * components: MuiGridList
   */
  content: PropTypes.arrayOf(PropTypes.shape({
    component: PropTypes.string.isRequired,
  })).isRequired,
};

MuiCollapse.defaultProps = {
  collapsedHeight: '0px',
  timeOut: 'auto',
  transitionIn: true,
  rootClass: [],
};
