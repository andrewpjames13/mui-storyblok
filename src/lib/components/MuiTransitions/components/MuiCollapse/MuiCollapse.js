import React from 'react';
import PropTypes from 'prop-types';
import { Collapse } from '@material-ui/core';
import useTransitionIn from '../../customHooks/useTransitionIn';
import StoryBlok from '../../../../utils/Storyblok';
import Blok from '../../../StoryBlokPage/components/Blok/Blok';
import { validComponentsRequired } from '../../../../utils/customProps';

const MuiCollapse = ({
  rootClass,
  collapsedHeight,
  transitionIn,
  enter,
  content,
  exit,
}) => {
  const styles = StoryBlok.arrayToMuiStyles(rootClass);
  const transTime = useTransitionIn(transitionIn);

  let enterTime;
  if (enter === '0') {
    enterTime = 'auto';
  } else {
    enterTime = { enter: +enter, exit: +exit };
  }

  return (
    <div className={styles.root}>
      <Collapse
        collapsedHeight={collapsedHeight}
        in={transTime}
        timeout={enterTime}
      >
        {content.map((item, index) => (
          <Blok {...item} key={index} />
        ))}
      </Collapse>
    </div>
  );
};

export default MuiCollapse;

MuiCollapse.propTypes = {
  /**
   * storyyblok multiselect of css classes
   * Mui Override or extend the styles applied to the component.
   */
  rootClass: PropTypes.arrayOf(PropTypes.string),
  /**
   * mui props: number | string
   * The height of the container when collapsed.
   */
  collapsedHeight: PropTypes.string,
  /**
   * mui props: number
   * Duration in milliseconds to enter the screen.
   */
  enter: PropTypes.string,
  /**
   * mui props: number
   * Duration in milliseconds to leave the screen.
   */
  exit: PropTypes.string,
  /**
   * mui props: number
   * Duration in milliseconds to enter the screen.
   * Set to 0 to automatically calculate transition time based on height.
   */
  transitionIn: PropTypes.string,
  /**
   * Content passed to render
   * components: Blok
   */
  content(props, propName, componentName) {
    return validComponentsRequired(props, propName, componentName, ['Blok'], 1);
  },
};

MuiCollapse.defaultProps = {
  collapsedHeight: '0px',
  transitionIn: 300,
  enter: 1500,
  exit: 1500,
  rootClass: [],
  content: [],
};
