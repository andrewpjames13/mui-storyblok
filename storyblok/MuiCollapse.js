module.exports = {
  "name": "MuiCollapse",
  "display_name": null,
  "updated_at": "2020-05-21T19:04:31.582Z",
  "schema": {
    "rootClass": {
      "type": "options",
      "description": "* stroyblok multiselect of css classes\n* Mui Override or extend the styles applied to the component."
    },
    "collapsedHeight": {
      "type": "text",
      "description": "* mui props: number | string\n* The height of the container when collapsed.",
      "default_value": "0px"
    },
    "content": {
      "type": "bloks",
      "maximum": "",
      "restrict_components": true,
      "component_whitelist": [
        "Blok"
      ],
      "required": true,
      "description": ""
    },
    "transitionIn": {
      "type": "number",
      "description": "Number of milliseconds for component to transition in.",
      "default_value": "300"
    },
    "enter": {
      "type": "number",
      "description": "* Duration in milliseconds to enter the screen.\n* Set to 0 to automatically calculate transition time based on height.",
      "display_name": "",
      "default_value": "1500"
    },
    "exit": {
      "type": "number",
      "description": "* mui props: number\n* Duration in milliseconds to leave the screen.",
      "display_name": "",
      "default_value": "1500"
    }
  },
  "image": null,
  "preview_field": null,
  "is_root": false,
  "preview_tmpl": null,
  "is_nestable": true,
  "all_presets": [],
  "preset_id": null,
  "real_name": "MuiCollapse",
  "component_group_uuid": null
}