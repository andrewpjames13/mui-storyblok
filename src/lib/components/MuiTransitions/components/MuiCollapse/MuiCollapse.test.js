import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import MuiCollapse from './MuiCollapse';

function setup(enterTime = 1500) {
  const props = {
    enter: enterTime,
    exit: 1500,
    content: [{
      component: 'Blok',
      content: [{
        component: 'MuiGrid',
        content: [{
          component: 'MuiGridItem',
          content: [{
            component: 'MuiCard',
            content: [{
              component: 'MuiCardContent',
              content: [{
                component: 'MuiTypography',
                content: [{
                  component: 'MuiText',
                  text: 'Juice Box',
                }],
              }],
            }],
          }],
        }],
      }],
    }],
  };

  const comp = mount(<MuiCollapse {...props} />);
  return { comp, props };
}

describe('<MuiCollapse />', () => {
  test('snapshot', () => {
    const { props } = setup();
    const tree = renderer.create(<MuiCollapse {...props} />);
    expect(tree).toMatchSnapshot();
  });

  it('should render MuiGridList', () => {
    const { comp } = setup();
    expect(comp).toMatchSnapshot();
  });

  it('should set enterTime to auto if enter value is 0', () => {
    const { comp } = setup('0');
    expect(comp.find('WithStyles(ForwardRef(Collapse))').first().props().timeout).toEqual('auto');
  });

  it('should change state after transitionIn time is ran', () => {
    const { comp } = setup();
    expect(comp.find('WithStyles(ForwardRef(Collapse))').first().props().in).toEqual(false);
    setTimeout(() => {
      expect(comp.find('WithStyles(ForwardRef(Collapse))').first().props().in).toEqual(true);
    }, 1000);
  });
});
