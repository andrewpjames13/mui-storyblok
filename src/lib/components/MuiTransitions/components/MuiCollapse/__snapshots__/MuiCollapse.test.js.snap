// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MuiCollapse /> should render MuiGridList 1`] = `ReactWrapper {}`;

exports[`<MuiCollapse /> snapshot 1`] = `
<div
  className="makeStyles-root-2"
>
  <div
    className="MuiCollapse-container MuiCollapse-hidden"
    style={
      Object {
        "minHeight": "0px",
      }
    }
  >
    <div
      className="MuiCollapse-wrapper"
    >
      <div
        className="MuiCollapse-wrapperInner"
      />
    </div>
  </div>
</div>
`;
